### PR TITLE
Sync with language-javascript

### LINKED ISSUE #18: Functions without arguments wrapped in parentheses won't be seen as functions.
A function without arguments wrapped in parentheses will not be processed as a function.

This is a function: 

``` coffee
dimensionNames: ( () ->
    ....
).property('dimensions.@each')
```

This isn't:

``` coffee
dimensionNames: ( ->
    ....
).property('dimensions.@each')
```

### LINKED ISSUE #25: Highlighting functions with multi line arguments not working
Consider this:

``` coffee
module.exports=
class Commit extends Treeish

  initialize: (
    ref
    repo
    @tree
    @parents
    @author
    @authoredDate
    @committer
    @committedDate
    @gpgsig
    @message
  ) -> super(ref, repo)

  @parse: (raw, repo) ->
    throw new Error('No valid git repo!!!') unless repo?.isGitRepo
    throw new Error('No raw data!!!') unless (typeof(raw) is 'string')
```

Your page is rendering it correctly but atom is not:

yousee

### LINKED ISSUE #45: Arguments where default values are function call result are highlighted incorrectly.
I’m guessing there’s a non-greedy match going on here, as only the second parameter is highlighted when everything from `(`up to `-` ought to be.

screen shot 2015-01-19 at 15 31 19

``` coffee
foo: (bar, baz = bob()) ->
```

### LINKED ISSUE #53: Many symbols without scopes
Most of the issues are in the way of parsing variable names. 

Load this example into atom. When I say something is `not tagged` I mean its scope is only `.source.coffee`.

``` coffee

someFunc = (arg1, arg2, arg3=1, arg4...) ->
  aVariable = 'ok'
  anArray = new Array()
  anArray.slice()           # anArray not tagged
  anArray.otherMethod()     # anArray, otherMethod not tagged
  randomGlobal.prop.omg()   # not tagged

  functionCall(one)         # functionCall not tagged
  functionCall one          # functionCall not tagged

  anObject =
    key1: 'val1'
    key2: (one, two) ->

# Totally messes this line up
grade = (student, period=(if b? then 7 else 6), messages={"A": "Excellent"}) ->
  if student.excellentWork  # not tagged
    "A+"
  else if student.okayStuff # not tagged
    if student.triedHard then "B" else "B-" # student.triedHard not tagged

class Animal extends Being
  constructor: (@name) ->   # @name not tagged the same as @name in the body, should be?
    @name += ' is cool'

# Ok
math =
  root:   Math.sqrt
  square: square
  cube:   (x) -> x * square x

# Ok
heredoc = """
  CoffeeScript subst test #{ 010 + 0xf / 0b10 + "nested string #{ /\n/ }"}
"""
```

Giving these symbols correct scopes will greatly improve the support in the coming `SymbolProvider` in the soon-to-be-bundled-in-core autocomplete-plus.

### LINKED ISSUE #63: Wrong `variable.parameter` matching
`variable.parameter` should match **`param1`** , **`param2`** 
Now: **`(param1, param2)`**

``` coffee
f1 = (param1, param2) -> x * x
```

### LINKED ISSUE #79: Wrong syntax highlighting in coffee for inline function without parameterlist
Syntax Highlighting seems off here, as these expressions are equal.
I build atom directly from the repository.
Well a fork actually, created 4th of december.
Couldn't find this in issues for this filter: is:issue is:open syntax Highlighting coffee 

Example:

``` coffee
# correct, this has class inline function
obj.method () ->
# wrong, class is entity name function coffee, but this is a call
obj.method ->

# both lines are identical and valid,
#  the second one is actually more idomatic
```

bildschirmfoto 2015-12-08 um 20 24 56

### LINKED ISSUE #88: Not highlighting functions 2
**Sticky** is a function but it's not highlighted

image

### LINKED ISSUE #90: Function name sometimes not highlighted correctly
If the `@` is removed, the function name is highlighted correctly.

``` coffee
obj.func @test
```

### LINKED ISSUE #97: Syntax highlight issue with destructuring expressions within string interpolations
screenshot at 00-28-59

### LINKED ISSUE #111: Should Strings as Argument Default Values be Highlighted?
In CoffeeScript

screenshot from 2017-01-17 10 50 09
```coffee
class Person
  greet: (msg='hello') ->
    console.log msg
```

`'hello'` is **not highlighted**, while in js:
```javascript
// JavaScript
function greet(msg='hello') {
  console.log(msg);
}
var greet = (msg='hello') =>
  console.log(msg);
```

Versions:
`language-coffee-script` v0.48.2
`Atom` v1.14.0-beta1

BTW I am not sure whether this issue is included in #103

### PR DESCRIPTION
**Pull request by 50Wliu**
This is a pretty large overhaul of language-coffee-script that aims to bring it up-to-date with language-javascript.  Things I've ported over include:
- Functions
- Function calls
- Method calls
- Objects
- Properties
- Numbers
- Operators
- Begin/end for braces

|Before|After|
|-------|-----|
|coffee-script-before|coffee-script-after|

Fixes #18 (they should both actually be variable assignments and not functions)
~~Fixes #25~~
Fixes #45
Fixes #63
Fixes #79
Fixes #88
Fixes #90
Fixes #97
Fixes #111
Closes #53

## TODO:
* [x] Instance variables aren't being tokenized correctly in parameters (see the fixme in the code).
* [ ] ~~Argument matching terminates at the end of the line instead of attempting to continue in the case of inline functions~~ - Unfixable using regex
* [ ] ~~Similar to above, argument matching does not continue for multiline arguments, even if they're surrounded by parentheses~~ - Unfixable using regex
* [x] Quoted functions are not recognized (`"test": ->`)
* [x] Objects turn into properties if you add an existence check (`a.b?.c`)
* [x] Properties don't turn into method calls if you add a standalone @ (`console.log @`)

/cc @atom/feedback - if some of you could try using this branch for everyday coding, that would be great in order to spot any regressions.
/cc @esdoppio @MaximSokolov @pchaigno

### CODE REVIEW

**Comment by 50Wliu** [comment]
Some things that need to be worked out:
* Function parameters are getting matched recursively when they shouldn't be.  An example is `grade = student((a)) ->`.  The inner parentheses shouldn't be getting `punctuation.definition.parameters.begin/end.bracket.round.coffee`.
* Regarding #18, the inner parentheses are directly being tokenizes as variables, which is wrong.
* Instance variables aren't being tokenized correctly in parameters (see the fixme in the code).

**Comment by 50Wliu** [comment]
Other than the "messed up" line in #53, everything else is now tagged.  After this is merged I'll open a new issue for that.

**Comment by pchaigno** [comment]
It's pretty great overall!
I found a couple constructions that seem broken by this:

0. [Example before (master) in Lightshow](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_url=https%3A%2F%2Fgithub.com%2Fatom%2Flanguage-coffee-script%2Fblob%2Fmaster%2Fgrammars%2Fcoffeescript.cson&grammar_text=&code_source=from-text&code_url=https%3A%2F%2Fgithub.com%2Fgithub%2Flinguist%2Fblob%2Fmaster%2Fsamples%2FCoffeeScript%2Flexer.coffee&code=++%23+Expand+variables+and+expressions+inside+double-quoted+strings+using%0D%0A++%23+Ruby-like+notation+for+substitution+of+arbitrary+expressions.%0D%0A++%23%0D%0A++%23+++++%22Hello+%23%7Bname.capitalize%28%29%7D.%22%0D%0A++%23%0D%0A++%23+If+it+encounters+an+interpolation%2C+this+method+will+recursively+create+a%0D%0A++%23+new+Lexer%2C+tokenize+the+interpolated+contents%2C+and+merge+them+into+the%0D%0A++%23+token+stream.%0D%0A++interpolateString%3A+%28str%2C+options+%3D+%7B%7D%29+-%3E%0D%0A++++%7Bheredoc%2C+regex%7D+%3D+options%0D%0A++++tokens+%3D+%5B%5D%0D%0A++++pi+%3D+0%0D%0A++++i++%3D+-1%0D%0A++++while+letter+%3D+str.charAt+i+%2B%3D+1%0D%0A++++++if+letter+is+%27%5C%5C%27%0D%0A++++++++i+%2B%3D+1%0D%0A++++++++continue%0D%0A++++++unless+letter+is+%27%23%27+and+str.charAt%28i%2B1%29+is+%27%7B%27+and%0D%0A+++++++++++++%28expr+%3D+%40balancedString+str%5Bi+%2B+1..%5D%2C+%27%7D%27%29%0D%0A++++++++continue%0D%0A++++++tokens.push+%5B%27NEOSTRING%27%2C+str%5Bpi...i%5D%5D+if+pi+%3C+i%0D%0A++++++inner+%3D+expr%5B1...-1%5D%0D%0A++++++if+inner.length%0D%0A++++++++nested+%3D+new+Lexer%28%29.tokenize+inner%2C+line%3A+%40line%2C+rewrite%3A+off%0D%0A++++++++nested.pop%28%29%0D%0A++++++++nested.shift%28%29+if+nested%5B0%5D%3F%5B0%5D+is+%27TERMINATOR%27%0D%0A++++++++if+len+%3D+nested.length%0D%0A++++++++++if+len+%3E+1%0D%0A++++++++++++nested.unshift+%5B%27%28%27%2C+%27%28%27%2C+%40line%5D%0D%0A++++++++++++nested.push++++%5B%27%29%27%2C+%27%29%27%2C+%40line%5D%0D%0A++++++++++tokens.push+%5B%27TOKENS%27%2C+nested%5D%0D%0A++++++i+%2B%3D+expr.length%0D%0A++++++pi+%3D+i+%2B+1%0D%0A++++tokens.push+%5B%27NEOSTRING%27%2C+str%5Bpi..%5D%5D+if+i+%3E+pi+%3C+str.length%0D%0A++++return+tokens+if+regex%0D%0A++++return+%40token+%27STRING%27%2C+%27%22%22%27+unless+tokens.length%0D%0A++++tokens.unshift+%5B%27%27%2C+%27%27%5D+unless+tokens%5B0%5D%5B0%5D+is+%27NEOSTRING%27%0D%0A++++%40token+%27%28%27%2C+%27%28%27+if+interpolated+%3D+tokens.length+%3E+1%0D%0A++++for+%5Btag%2C+value%5D%2C+i+in+tokens%0D%0A++++++%40token+%27%2B%27%2C+%27%2B%27+if+i%0D%0A++++++if+tag+is+%27TOKENS%27%0D%0A++++++++%40tokens.push+value...%0D%0A++++++else%0D%0A++++++++%40token+%27STRING%27%2C+%40makeString+value%2C+%27%22%27%2C+heredoc%0D%0A++++%40token+%27%29%27%2C+%27%29%27+if+interpolated%0D%0A++++tokens)
0. [Example with this pull request](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_url=https%3A%2F%2Fgithub.com%2Fatom%2Flanguage-coffee-script%2Fblob%2F80bc586cede3c87dbb81e827d90c13dfb7c9345f%2Fgrammars%2Fcoffeescript.cson&grammar_text=&code_source=from-text&code_url=https%3A%2F%2Fgithub.com%2Fgithub%2Flinguist%2Fblob%2Fmaster%2Fsamples%2FCoffeeScript%2Flexer.coffee&code=++%23+Expand+variables+and+expressions+inside+double-quoted+strings+using%0D%0A++%23+Ruby-like+notation+for+substitution+of+arbitrary+expressions.%0D%0A++%23%0D%0A++%23+++++%22Hello+%23%7Bname.capitalize%28%29%7D.%22%0D%0A++%23%0D%0A++%23+If+it+encounters+an+interpolation%2C+this+method+will+recursively+create+a%0D%0A++%23+new+Lexer%2C+tokenize+the+interpolated+contents%2C+and+merge+them+into+the%0D%0A++%23+token+stream.%0D%0A++interpolateString%3A+%28str%2C+options+%3D+%7B%7D%29+-%3E%0D%0A++++%7Bheredoc%2C+regex%7D+%3D+options%0D%0A++++tokens+%3D+%5B%5D%0D%0A++++pi+%3D+0%0D%0A++++i++%3D+-1%0D%0A++++while+letter+%3D+str.charAt+i+%2B%3D+1%0D%0A++++++if+letter+is+%27%5C%5C%27%0D%0A++++++++i+%2B%3D+1%0D%0A++++++++continue%0D%0A++++++unless+letter+is+%27%23%27+and+str.charAt%28i%2B1%29+is+%27%7B%27+and%0D%0A+++++++++++++%28expr+%3D+%40balancedString+str%5Bi+%2B+1..%5D%2C+%27%7D%27%29%0D%0A++++++++continue%0D%0A++++++tokens.push+%5B%27NEOSTRING%27%2C+str%5Bpi...i%5D%5D+if+pi+%3C+i%0D%0A++++++inner+%3D+expr%5B1...-1%5D%0D%0A++++++if+inner.length%0D%0A++++++++nested+%3D+new+Lexer%28%29.tokenize+inner%2C+line%3A+%40line%2C+rewrite%3A+off%0D%0A++++++++nested.pop%28%29%0D%0A++++++++nested.shift%28%29+if+nested%5B0%5D%3F%5B0%5D+is+%27TERMINATOR%27%0D%0A++++++++if+len+%3D+nested.length%0D%0A++++++++++if+len+%3E+1%0D%0A++++++++++++nested.unshift+%5B%27%28%27%2C+%27%28%27%2C+%40line%5D%0D%0A++++++++++++nested.push++++%5B%27%29%27%2C+%27%29%27%2C+%40line%5D%0D%0A++++++++++tokens.push+%5B%27TOKENS%27%2C+nested%5D%0D%0A++++++i+%2B%3D+expr.length%0D%0A++++++pi+%3D+i+%2B+1%0D%0A++++tokens.push+%5B%27NEOSTRING%27%2C+str%5Bpi..%5D%5D+if+i+%3E+pi+%3C+str.length%0D%0A++++return+tokens+if+regex%0D%0A++++return+%40token+%27STRING%27%2C+%27%22%22%27+unless+tokens.length%0D%0A++++tokens.unshift+%5B%27%27%2C+%27%27%5D+unless+tokens%5B0%5D%5B0%5D+is+%27NEOSTRING%27%0D%0A++++%40token+%27%28%27%2C+%27%28%27+if+interpolated+%3D+tokens.length+%3E+1%0D%0A++++for+%5Btag%2C+value%5D%2C+i+in+tokens%0D%0A++++++%40token+%27%2B%27%2C+%27%2B%27+if+i%0D%0A++++++if+tag+is+%27TOKENS%27%0D%0A++++++++%40tokens.push+value...%0D%0A++++++else%0D%0A++++++++%40token+%27STRING%27%2C+%40makeString+value%2C+%27%22%27%2C+heredoc%0D%0A++++%40token+%27%29%27%2C+%27%29%27+if+interpolated%0D%0A++++tokens)

Notice a few keywords are not highlighted near the end.

**Comment by 50Wliu** [comment]
Thanks - fixed!
[Lightshow](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_url=https%3A%2F%2Fraw.githubusercontent.com%2Fatom%2Flanguage-coffee-script%2Fbdbf876dd9162d3c3854da85f4d2dbc037964b67%2Fgrammars%2Fcoffeescript.cson&grammar_text=&code_source=from-text&code_url=https%3A%2F%2Fgithub.com%2Fgithub%2Flinguist%2Fblob%2Fmaster%2Fsamples%2FCoffeeScript%2Flexer.coffee&code=++%23+Expand+variables+and+expressions+inside+double-quoted+strings+using%0D%0A++%23+Ruby-like+notation+for+substitution+of+arbitrary+expressions.%0D%0A++%23%0D%0A++%23+++++%22Hello+%23%7Bname.capitalize%28%29%7D.%22%0D%0A++%23%0D%0A++%23+If+it+encounters+an+interpolation%2C+this+method+will+recursively+create+a%0D%0A++%23+new+Lexer%2C+tokenize+the+interpolated+contents%2C+and+merge+them+into+the%0D%0A++%23+token+stream.%0D%0A++interpolateString%3A+%28str%2C+options+%3D+%7B%7D%29+-%3E%0D%0A++++%7Bheredoc%2C+regex%7D+%3D+options%0D%0A++++tokens+%3D+%5B%5D%0D%0A++++pi+%3D+0%0D%0A++++i++%3D+-1%0D%0A++++while+letter+%3D+str.charAt+i+%2B%3D+1%0D%0A++++++if+letter+is+%27%5C%5C%27%0D%0A++++++++i+%2B%3D+1%0D%0A++++++++continue%0D%0A++++++unless+letter+is+%27%23%27+and+str.charAt%28i%2B1%29+is+%27%7B%27+and%0D%0A+++++++++++++%28expr+%3D+%40balancedString+str%5Bi+%2B+1..%5D%2C+%27%7D%27%29%0D%0A++++++++continue%0D%0A++++++tokens.push+%5B%27NEOSTRING%27%2C+str%5Bpi...i%5D%5D+if+pi+%3C+i%0D%0A++++++inner+%3D+expr%5B1...-1%5D%0D%0A++++++if+inner.length%0D%0A++++++++nested+%3D+new+Lexer%28%29.tokenize+inner%2C+line%3A+%40line%2C+rewrite%3A+off%0D%0A++++++++nested.pop%28%29%0D%0A++++++++nested.shift%28%29+if+nested%5B0%5D%3F%5B0%5D+is+%27TERMINATOR%27%0D%0A++++++++if+len+%3D+nested.length%0D%0A++++++++++if+len+%3E+1%0D%0A++++++++++++nested.unshift+%5B%27%28%27%2C+%27%28%27%2C+%40line%5D%0D%0A++++++++++++nested.push++++%5B%27%29%27%2C+%27%29%27%2C+%40line%5D%0D%0A++++++++++tokens.push+%5B%27TOKENS%27%2C+nested%5D%0D%0A++++++i+%2B%3D+expr.length%0D%0A++++++pi+%3D+i+%2B+1%0D%0A++++tokens.push+%5B%27NEOSTRING%27%2C+str%5Bpi..%5D%5D+if+i+%3E+pi+%3C+str.length%0D%0A++++return+tokens+if+regex%0D%0A++++return+%40token+%27STRING%27%2C+%27%22%22%27+unless+tokens.length%0D%0A++++tokens.unshift+%5B%27%27%2C+%27%27%5D+unless+tokens%5B0%5D%5B0%5D+is+%27NEOSTRING%27%0D%0A++++%40token+%27%28%27%2C+%27%28%27+if+interpolated+%3D+tokens.length+%3E+1%0D%0A++++for+%5Btag%2C+value%5D%2C+i+in+tokens%0D%0A++++++%40token+%27%2B%27%2C+%27%2B%27+if+i%0D%0A++++++if+tag+is+%27TOKENS%27%0D%0A++++++++%40tokens.push+value...%0D%0A++++++else%0D%0A++++++++%40token+%27STRING%27%2C+%40makeString+value%2C+%27%22%27%2C+heredoc%0D%0A++++%40token+%27%29%27%2C+%27%29%27+if+interpolated%0D%0A++++tokens)

File: grammars/coffeescript.cson [hunk 1]
```diff
@@ -393,6 +448,75 @@
...
+        'patterns': [
+          {
+            'include': '$self'
+          }
+          {
+            'match': '(@)?[a-zA-Z_$][\\w$]*(\\.\\.\\.)?'
+            'captures':
+              '0':
+                'name': 'variable.parameter.function.coffee'
+              '1': # FIXME: Doesn't work yet: the instance_variable include overrides this
```
Comment (MaximSokolov) [inline grammars/coffeescript.cson]: It probably should be
```coffee
{
  'match': '([a-zA-Z_$][\\w$]*)(\\.\\.\\.)?'
  'captures':
    '1':
      'name': 'variable.parameter.function.coffee'
    '2':
      'name': 'keyword.operator.splat.coffee'
}
{
  'match': '(@(?:[a-zA-Z_$][\\w$]*)?)(\\.\\.\\.)?'
  'captures':
    '1':
      'name': 'variable.parameter.function.readwrite.instance.coffee'
    '2':
      'name': 'keyword.operator.splat.coffee'
}
{
  'include': '$self'
}
```

File: grammars/coffeescript.cson [hunk 2]
```diff
@@ -167,7 +157,49 @@
...
+      \\s*([:=])\\s*
+      (?=(\\([^\\(\\)]*\\))?\\s*[=-]>)
+    '''
+    'beginCaptures':
+      '1':
+        'name': 'entity.name.function.coffee'
+      '2':
+        'name': 'variable.other.readwrite.instance.coffee'
+      '3':
+        'name': 'keyword.operator.assignment.coffee'
```
Comment (MaximSokolov) [inline grammars/coffeescript.cson]: Note that `=`, `:` operators now scoped differently
```coffee
test = () -> x * x # keyword.operator.assignment
a = 1 + 1 # just keyword.operator
```
> image
Seti syntax

**Comment by MaximSokolov** [comment]
It also closes #64

**Comment by 50Wliu** [comment]
To be clear, this suffers from the same problems in #64 as language-javascript also can't tokenize those functions.

**Comment by 50Wliu** [comment]
Thanks @MaximSokolov - your fix did work!

**Comment by 50Wliu** [comment]
I think this is ready for review.

**Comment by lee-dohm** [comment]
I've been running this for a couple days and things look good to me :+1:

**Comment by 50Wliu** [comment]
```coffee
console.log not hasScope(beforePrefixScopesArray, "entity.name.tag.css") and prefix.trim() is ":"
```
^ reminder for me to fix that

**Comment by 50Wliu** [comment]
`stack` is incorrectly tokenized as a function:
```coffee
for currentStack in stack by -1
```

No support for `..`:
```coffee
index = parseInt(match[1..])
```

**Comment by 50Wliu** [comment]
```coffee
error = new Error("Deprecated function(s) #{deprecations.map(({originName}) -> originName).join ', '}) were called.")
```

**Comment by Ben3eeE** [comment]
:tada:

**Pull request merged**